app/feature/feature-travel-certificate/src/main/kotlin/com/hedvig/android/feature/travelcertificate/ui/generate_when/TravelCertificateDateInputViewModel.kt

```diff
@@ -0,0 +1,231 @@
+import kotlinx.datetime.LocalDate
+import kotlinx.datetime.TimeZone
+import kotlinx.datetime.toLocalDateTime
+import kotlinx.serialization.Serializable
+
+internal class TravelCertificateDateInputViewModel(
+  contractId: String?,
+  getTravelCertificateSpecificationsUseCase: GetTravelCertificateSpecificationsUseCase,
+  createTravelCertificateUseCase: CreateTravelCertificateUseCase,
+) : MoleculeViewModel<TravelCertificateDateInputEvent, TravelCertificateDateInputUiState>(
+    initialState = TravelCertificateDateInputUiState.Loading,
+    presenter = TravelCertificateDateInputPresenter(
+      contractId,
+      getTravelCertificateSpecificationsUseCase,
+      createTravelCertificateUseCase,
+    ),
+  )
+
+internal class TravelCertificateDateInputPresenter(
+  private val contractId: String?,
+  private val getTravelCertificateSpecificationsUseCase: GetTravelCertificateSpecificationsUseCase,
+  private val createTravelCertificateUseCase: CreateTravelCertificateUseCase,
+) : MoleculePresenter<TravelCertificateDateInputEvent, TravelCertificateDateInputUiState> {
+  @Composable
+  override fun MoleculePresenterScope<TravelCertificateDateInputEvent>.present(
+    lastState: TravelCertificateDateInputUiState,
+  ): TravelCertificateDateInputUiState {
+    var loadIteration by remember { mutableIntStateOf(0) }
+
+    var generateIteration by remember { mutableIntStateOf(0) }
+
+    var screenContent by remember { mutableStateOf<DateInputScreenContent>(DateInputScreenContent.Loading) }
+
+    var travelDate by remember {
+      mutableStateOf(
+        Clock.System.now().toLocalDateTime(TimeZone.currentSystemDefault()).date,
+      )
+    }
+
+    var email by remember { mutableStateOf<ValidatedInput<String?>>(ValidatedInput(null)) }
+
+    CollectEvents { event ->
+      when (event) {
+        TravelCertificateDateInputEvent.RetryLoadData -> {
+          loadIteration++
+        }
+
+        is TravelCertificateDateInputEvent.ChangeEmailInput -> {
+          email = ValidatedInput(event.email)
+        }
+
+        is TravelCertificateDateInputEvent.ChangeDataInput -> {
+          travelDate = event.localDate
+        }
+
+        TravelCertificateDateInputEvent.GenerateTravelCertificate -> {
+          generateIteration++
+        }
+      }
+    }
+
+    LaunchedEffect(generateIteration) {
+      val currentContent = screenContent
+      val currentEmail = email.input
+      if (currentContent is DateInputScreenContent.Success && currentEmail != null) {
+        screenContent = DateInputScreenContent.Loading
+        createTravelCertificateUseCase.invoke(
+          contractId = currentContent.details.contractId,
+          startDate = travelDate,
+          isMemberIncluded = true,
+          coInsured = listOf(),
+          email = currentEmail,
+        ).fold(
+          ifLeft = { _ ->
+            screenContent = DateInputScreenContent.Failure
+            // todo: so on Retry we would just show the initial DateInput screen again
```

Comment:
I like to have a `errorMessage: String?` in the uiState, which you can set here. Then you can show a dialog in the view when this errorMessage is non-null. Let the dialog call the viewmodel when dismissed to set the errorMessage to null again.